// notasrest/notasrest.js
var rest = require('restler');
var fs = require('fs');
var netconfig = require('netconfig');
var host = netconfig.getHost();
var port = netconfig.getPort();

module.exports = {

   getNotas: function(token,callback){
            rest.get('http://'+host+':'+port+'/notas/notas/all',{
               accessToken: token
            })
            .on('success', function(data, response){
                 callback(data);
            })
            .on('error', function(err, response){
                 callback({message:"Erro ao buscar todas as Notas"});
            });
   },

   getFirstNotas: function(token,callback){
            rest.get("http://"+host+":"+port+"/notas/notas/first/20",{
                 accessToken: token
            })
            .on('success', function(data, response){
                 callback(data);
            })
            .on('error', function(err, response){
                 callback({message:"Erro ao buscar todas as Notas"});
            });
    },

   newNota: function(token,notadata, callback){
            rest.post("http://"+host+":"+port+"/notas/notas/new", {
               accessToken: token,
               data: notadata
            })
            .on('success', function(data, response){
               callback(data);
            })
            .on('error', function(err, response){
               callback({message:"Erro ao buscar todas as Notas"});
            });
   },

   getNotasLike: function(token, parameters, callback){
     if (parameters !=""){
                 parameters = encodeURIComponent(parameters);
                 rest.get("http://"+host+":"+port+"/notas/notas/like/"+parameters,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao pesquisar uma Nota!!"});
                 });
     }
   },

   deleteNotaById: function(token, id, callback){
      if (id!=""){
                 rest.del("http://"+host+":"+port+"/notas/notas/id/"+id,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao apagar uma nota!!"});
                 });
      }
   },

   deleteNotaByCodigo: function(token, codigo,callback){
      if (codigo!="") {
                 rest.del("http://"+host+":"+port+"/notas/notas/codigo/"+codigo,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao apagar uma nota!!"});
                 });
      }
   },

   updateNotaByCodigo: function(token, codigo, notadata, callback){
      if (codigo!="") {
                 notadata.nota = encodeURIComponent(notadata.nota);
                 notadata.tags = encodeURIComponent(notadata.tags);
                 rest.put("http://"+host+":"+port+"/notas/notas/codigo/"+codigo,{
                       multipart: true,
                       accessToken: token,
                       data: notadata
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao atualizar Nota no banco de dados!!"});
                 });
      } else {
           callback({message:"Código não informado!!"});
      }
   },

   updateNotaById: function(token, id, notadata,callback){
      if (id!=""){
                 rest.put("http://"+host+":"+port+"/notas/notas/id/"+id,{
                    accessToken: token,
                    data: notadata
                  })
                  .on('success', function(data,response){
                      callback(data);
                  })
                  .on('error', function(err, response){
                      callback({message:"Erro ao atualizar Nota no banco de dados!!"});
                  });
     } else {
          callback({message:"Id não informado!!"});
     }
   },

   getNotaByCodigoLike: function(token, codigo,callback){
      if (codigo!=""){
                 rest.get("http://"+host+":"+port+"/notas/notas/codigo/like/"+codigo,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
      } else {
          callback({message: "Código não informado!!"});
      }
   },

   getNotaByCodigo: function(token, codigo, callback){
      if (codigo!=""){
                 rest.get("http://"+host+":"+port+"/notas/notas/codigo/"+codigo,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
      } else {
          callback({message: "Código não informado!!"});
      }
   },

   getNotaById: function(token, id, callback){
      if (id!=""){
                 rest.get("http://"+host+":"+port+"/notas/notas/id/"+id,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
      } else {
          callback({message:"Identificador não informado!!"});
      }
   },

   getNotasByTagsOr: function(token, tags, callback){
      var vector = tags.split(",");

      //Coleta todos os campos da busca por tag.
      var parameters = "";
      if (vector instanceof Array){
            if (vector[0] != "") {
                   parameters=parameters+"tags="+encodeURIComponent(vector[0].trim());
                   for (var i=1;i<vector.length;i++) {
                      parameters=parameters+"&tags="+encodeURIComponent(vector[i].trim());
                   }
            }
      }

      if (parameters!="") {
                 parameters="?"+parameters;
                 rest.get("http://"+host+":"+port+"/notas/notas/tags/or"+parameters,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
      } else {
             callback({message:"Parâmetros não foram definidos corretamente!!"});
      }
   },

   getNotasByTagsAnd: function(token, tags, callback){
       var vector = tags.split(",");

       //Coleta todos os campos da busca por tag.
       var parameters = "";
       if (vector instanceof Array){
             if (vector[0] != "") {
                   parameters=parameters+"tags="+encodeURIComponent(vector[0].trim());
                   for (var i=1;i<vector.length;i++) {
                          parameters=parameters+"&tags="+encodeURIComponent(vector[i].trim());
                   }
            }
      }

      if (parameters!="") {
                 parameters="?"+parameters;
                 rest.get("http://"+host+":"+port+"/notas/notas/tags/and"+parameters,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
      } else {
             callback({message:"Parâmetros não foram definidos corretamente!!"});
      }
   },

   getNotasByTags: function(token, searchTags, callback) {
      if (searchTags != "") {
                 searchTags = "?"+searchTags;
                 rest.get("http://"+host+":"+port+"/notas/notas/tags/search"+searchTags,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
     } else {
           callback({message: "Parâmetros não foram definidos corretamente!!"});
     }
   },

   getNotasByTagsLike: function(token, tags,callback){
       var vector = tags.split(",");

       //Coleta todos os campos da busca por tag.
       var parameters = "";
       if (vector instanceof Array){
          if (vector[0] != "") {
             parameters=parameters+"tags="+encodeURIComponent(vector[0].trim());
             for (var i=1;i<vector.length;i++) {
                parameters=parameters+"&tags="+encodeURIComponent(vector[i].trim());
             }
          }
       }

      if (parameters!="") {
                 parameters="?"+parameters;
                 rest.get("http://"+host+":"+port+"/notas/notas/tags/like"+parameters,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar Nota no banco de dados!!"});
                 });
       } else {
           callback({message:"Parâmetros não foram definidos corretamente!!"});
       }
   },

   getTagsMapReduce: function(token, callback){
           rest.get("http://"+host+":"+port+"/notas/mapReduce/Tags",{
                 accessToken: token
           })
           .on("success", function(data, response){
                 callback(data);
           })
           .on('error', function(err, response){
                 callback({message: "Erro ao buscar arquivo!!"});
           });
   },

   getDocumentInfo: function(token, codigo,doc,callback){
      if (doc !="" && codigo!=""){
                 rest.get("http://"+host+":"+port+"/notas/notas/"+codigo+"/arquivo/"+doc+'/info',{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar arquivo!!"});
                 });
      } else {
           callback({message: "Arquivo não foi encontrado!!"});
      }
   },

   getDocument: function(token, doc, callback){
      if (doc !=""){
                 rest.get("http://"+host+":"+port+"/arquivos/"+doc,{
                       accessToken: token
                 })
                 .on('success', function(data,response){
                       callback(data);
                 })
                 .on('error', function(err, response){
                       callback({message:"Erro ao buscar arquivo!!"});
                 });
      } else {
            callback({message: "Arquivo não foi encontrado!!"});
      }
   },

   insertDocument: function(doc,callback){
       //TODO - maybe it'll be implemented later
   },

   deleteDocument: function(token, codigo,doc,callback){
      if ((doc!="") && (codigo!="")) {
           fs.readFile('./userInfo','utf8', function(err,data){
                 var info = JSON.parse(data);
                 rest.del("http://"+host+":"+port+"/notas/notas/"+codigo+"/arquivo/"+doc,{
                     accessToken: JSON.parse(fs.readFileSync('./userInfo','utf8')).token
                 })
                 .on('success', function(data,response){
                     callback(data);
                 })
                 .on('error', function(err, response){
                     callback({message:"Erro ao apagar arquivo!!"});
                 });
           });
      } else {
            callback({message: "Arquivo não foi encontrado!!"});
      }
  }

};
